creating and removing connections eventually leads to error _unreserve_vlan Not implemented  AssertionError

Hi,

After creating and removing a few VLANs (one of them being the case reported on https://github.com/atlanticwave-sdx/sdx-controller/issues/309), eventually we get the following error:

```
2024-08-10T01:42:45.419357804Z INFO:sdx_pce.topology.temanager:reserve_vlan_breakdown: domain_breakdown: {'urn:sdx:topology:ampath.net': {'ingress_port': {'id': 'urn:sdx:port:ampath.net:Ampath3:50', 'name': 'Ampath3-eth50', 'short_name': None, 'node': 'urn:sdx:node:ampath.net:Ampath3', 'vlan_range': None, 'status': 'up', 'state': 'enabled', 'nni': '', 'type': '10GE', 'services': {'l2vpn_ptp': {'vlan_range': [[1, 4095]]}, 'l2vpn_ptmp': {}, 'monitoring_capability': None, 'owner': None, 'private_attributes': None, 'provisioning_system': None, 'provisioning_url': None, 'vendor': None}, 'private_attributes': None}, 'egress_port': {'id': 'urn:sdx:port:ampath.net:Ampath1:40', 'name': 'Ampath1-eth40', 'short_name': None, 'node': 'urn:sdx:node:ampath.net:Ampath1', 'vlan_range': None, 'status': 'up', 'state': 'enabled', 'nni': 'urn:sdx:port:sax.net:Sax01:40', 'type': '10GE', 'services': {'l2vpn_ptp': {'vlan_range': [[1, 4095]]}, 'l2vpn_ptmp': {}, 'monitoring_capability': None, 'owner': None, 'private_attributes': None, 'provisioning_system': None, 'provisioning_url': None, 'vendor': None}, 'private_attributes': None}}, 'urn:sdx:topology:sax.net': {'ingress_port': {'id': 'urn:sdx:port:sax.net:Sax01:40', 'name': 'Sax01-eth40', 'short_name': None, 'node': 'urn:sdx:node:sax.net:Sax01', 'vlan_range': None, 'status': 'up', 'state': 'enabled', 'nni': 'urn:sdx:port:ampath.net:Ampath1:40', 'type': '10GE', 'services': {'l2vpn_ptp': {'vlan_range': [[1, 4095]]}, 'l2vpn_ptmp': {}, 'monitoring_capability': None, 'owner': None, 'private_attributes': None, 'provisioning_system': None, 'provisioning_url': None, 'vendor': None}, 'private_attributes': None}, 'egress_port': {'id': 'urn:sdx:port:sax.net:Sax01:41', 'name': 'Sax01-eth41', 'short_name': None, 'node': 'urn:sdx:node:sax.net:Sax01', 'vlan_range': None, 'status': 'up', 'state': 'enabled', 'nni': 'urn:sdx:port:tenet.ac.za:Tenet01:41', 'type': '10GE', 'services': {'l2vpn_ptp': {'vlan_range': [[1, 4095]]}, 'l2vpn_ptmp': {}, 'monitoring_capability': None, 'owner': None, 'private_attributes': None, 'provisioning_system': None, 'provisioning_url': None, 'vendor': None}, 'private_attributes': None}}, 'urn:sdx:topology:tenet.ac.za': {'ingress_port': {'id': 'urn:sdx:port:tenet.ac.za:Tenet01:41', 'name': 'Tenet01-eth41', 'short_name': None, 'node': 'urn:sdx:node:tenet.ac.za:Tenet01', 'vlan_range': None, 'status': 'up', 'state': 'enabled', 'nni': 'urn:sdx:port:sax.net:Sax01:41', 'type': '10GE', 'services': {'l2vpn_ptp': {'vlan_range': [[1, 4095]]}, 'l2vpn_ptmp': {}, 'monitoring_capability': None, 'owner': None, 'private_attributes': None, 'provisioning_system': None, 'provisioning_url': None, 'vendor': None}, 'private_attributes': None}, 'egress_port': {'id': 'urn:sdx:port:tenet.ac.za:Tenet03:50', 'name': 'Tenet03-eth50', 'short_name': None, 'node': 'urn:sdx:node:tenet.ac.za:Tenet03', 'vlan_range': None, 'status': 'up', 'state': 'enabled', 'nni': '', 'type': '10GE', 'services': {'l2vpn_ptp': {'vlan_range': [[1, 4095]]}, 'l2vpn_ptmp': {}, 'monitoring_capability': None, 'owner': None, 'private_attributes': None, 'provisioning_system': None, 'provisioning_url': None, 'vendor': None}, 'private_attributes': None}}}
2024-08-10T01:42:45.421833645Z WARNING:sdx_pce.topology.temanager:Port urn:sdx:port:ampath.net:Ampath3:50 not found in port map, it's a user port, by default uses the OXP vlan
2024-08-10T01:42:45.421846288Z INFO:sdx_pce.topology.temanager:VLAN reservation: domain: urn:sdx:topology:ampath.net, ingress_vlan: 4094, egress_vlan: 4094
2024-08-10T01:42:45.425263310Z INFO:sdx_pce.topology.temanager:VLAN reservation: domain: urn:sdx:topology:sax.net, ingress_vlan: 4094, egress_vlan: 4094
2024-08-10T01:42:45.425351835Z WARNING:sdx_pce.topology.temanager:reserve_vlan domain: urn:sdx:topology:tenet.ac.za entry: None
2024-08-10T01:42:45.425380157Z WARNING:sdx_pce.topology.temanager:reserve_vlan domain: urn:sdx:topology:tenet.ac.za entry: None
2024-08-10T01:42:45.425406236Z WARNING:sdx_pce.topology.temanager:Port urn:sdx:port:tenet.ac.za:Tenet01:41 not found in port map, it's a user port, by default uses the OXP vlan
2024-08-10T01:42:45.425423154Z WARNING:sdx_pce.topology.temanager:Port urn:sdx:port:tenet.ac.za:Tenet03:50 not found in port map, it's a user port, by default uses the OXP vlan
2024-08-10T01:42:45.425427837Z INFO:sdx_pce.topology.temanager:VLAN reservation: domain: urn:sdx:topology:tenet.ac.za, ingress_vlan: None, egress_vlan: None
2024-08-10T01:42:45.425580351Z ERROR:sdx_controller.handlers.connection_handler:Error when generating/publishing breakdown: Not implemented - Traceback (most recent call last):,   File "/opt/venv/lib/python3.9/site-packages/sdx_controller/handlers/connection_handler.py", line 157, in place_connection,     breakdown = te_manager.generate_connection_breakdown(,   File "/opt/venv/lib/python3.9/site-packages/sdx_pce/topology/temanager.py", line 620, in generate_connection_breakdown,     tagged_breakdown = self._reserve_vlan_breakdown(,   File "/opt/venv/lib/python3.9/site-packages/sdx_pce/topology/temanager.py", line 792, in _reserve_vlan_breakdown,     self._unreserve_vlan(domain, ingress_port, ingress_vlan),   File "/opt/venv/lib/python3.9/site-packages/sdx_pce/topology/temanager.py", line 933, in _unreserve_vlan,     assert False, "Not implemented", AssertionError: Not implemented,
2024-08-10T01:42:45.425619108Z INFO:sdx_controller.controllers.connection_controller:place_connection result: ID: e186a378-0452-4cf3-89a2-bbf30a6ed38a reason='Error: Not implemented', code=400
2024-08-10T01:42:45.426459635Z INFO:     192.168.0.1:35116 - "POST /SDX-Controller/1.0.0/connection HTTP/1.1" 400 Bad Request
```

The request above was basically:

```
vlan1=301; vlan2=301; curl -s -X POST -H 'Content-type: application/json' http://0.0.0.0:8080/SDX-Controller/1.0.0/connection -d '{"name": "VLAN between AMPATH/'$vlan1' and TENET/'$vlan2'", "endpoints": [{"port_id": "urn:sdx:port:ampath.net:Ampath3:50", "vlan": "'$vlan1'"}, {"port_id": "urn:sdx:port:tenet.ac.za:Tenet03:50", "vlan": "'$vlan2'"}]}'
```

Hello, team. Did you guys have the opportunity to check this issue? Ccing @YufengXin @sajith and @congwang09 

I'm writing now because once again I was running tests and faced this issue. The SDX controller logs say this:

```
$ docker logs sdx-controller -t > logs-sdx-controller
$ tail -n 10 logs-sdx-controller
2024-08-28T09:08:01.282953671Z WARNING:sdx_pce.topology.temanager:Can't find a mapping for domain:urn:sdx:topology:tenet.ac.za port:urn:sdx:port:tenet.ac.za:Tenet03:50
2024-08-28T09:08:01.282997534Z WARNING:sdx_pce.topology.temanager:Port urn:sdx:port:tenet.ac.za:Tenet01:41 not found in port map, it's a user port, by default uses the OXP vlan
2024-08-28T09:08:01.283029531Z WARNING:sdx_pce.topology.temanager:Port urn:sdx:port:tenet.ac.za:Tenet03:50 not found in port map, it's a user port, by default uses the OXP vlan
2024-08-28T09:08:01.283043683Z INFO:sdx_pce.topology.temanager:VLAN reservation: domain: urn:sdx:topology:tenet.ac.za, ingress_vlan: None, egress_vlan: None
2024-08-28T09:08:01.283552127Z ERROR:sdx_controller.handlers.connection_handler:Error when generating/publishing breakdown: Not implemented - Traceback (most recent call last):,   File "/opt/venv/lib/python3.9/site-packages/sdx_controller/handlers/connection_handler.py", line 157, in place_connection,     breakdown = te_manager.generate_connection_breakdown(,   File "/opt/venv/lib/python3.9/site-packages/sdx_pce/topology/temanager.py", line 620, in generate_connection_breakdown,     tagged_breakdown = self._reserve_vlan_breakdown(,   File "/opt/venv/lib/python3.9/site-packages/sdx_pce/topology/temanager.py", line 792, in _reserve_vlan_breakdown,     self._unreserve_vlan(domain, ingress_port, ingress_vlan),   File "/opt/venv/lib/python3.9/site-packages/sdx_pce/topology/temanager.py", line 933, in _unreserve_vlan,     assert False, "Not implemented", AssertionError: Not implemented,
2024-08-28T09:08:01.283590780Z INFO:sdx_controller.controllers.l2vpn_controller:place_connection result: ID: 15e2dae0-34cc-4256-af5d-e178aa85e07e reason='Error: Not implemented', code=400
2024-08-28T09:08:01.283654623Z DEBUG:connexion.apis.abstract:Getting data and status code
2024-08-28T09:08:01.283825341Z DEBUG:connexion.apis.abstract:Prepared body and status code (400)
2024-08-28T09:08:01.283903271Z DEBUG:connexion.apis.abstract:Got framework response
2024-08-28T09:08:01.284573954Z INFO:     192.168.0.1:50760 - "POST /SDX-Controller/l2vpn/1.0 HTTP/1.1" 400 Bad Request
```

Which indicates the issue was missing vlan table on TENET domain.

Looking on how TENET domain had its topology imported, we have:

```
$ egrep 'MQ received message:.* "urn:sdx:topology:tenet.ac.za"' logs-sdx-controller
2024-08-28T09:02:10.937524682Z DEBUG:sdx_controller.messaging.rpc_queue_consumer:MQ received message:b'{"name": "Tenet-OXP", "id": "urn:sdx:topology:tenet.ac.za", "version": 3, "timestamp": "2024-08-28T09:02:07Z", "model_version": "2.0.0", "nodes": [{"name": "Tenet02", "id": "urn:sdx:node:tenet.ac.za:Tenet02", "location": {"address": "", "latitude": 0.0, "longitude": -30.0, "iso3166_2_lvl4": "", "private": []}, "ports": [{"id": "urn:sdx:port:tenet.ac.za:Tenet02:41", "name": "Tenet02-eth41", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet02:1", "name": "Tenet02-eth1", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet02:50", "name": "Tenet02-eth50", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}], "status": "up", "state": "enabled"}], "links": [], "services": ["l2vpn-ptp"]}'
2024-08-28T09:02:10.937726512Z INFO:sdx_controller.handlers.lc_message_handler:MQ received message:b'{"name": "Tenet-OXP", "id": "urn:sdx:topology:tenet.ac.za", "version": 3, "timestamp": "2024-08-28T09:02:07Z", "model_version": "2.0.0", "nodes": [{"name": "Tenet02", "id": "urn:sdx:node:tenet.ac.za:Tenet02", "location": {"address": "", "latitude": 0.0, "longitude": -30.0, "iso3166_2_lvl4": "", "private": []}, "ports": [{"id": "urn:sdx:port:tenet.ac.za:Tenet02:41", "name": "Tenet02-eth41", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet02:1", "name": "Tenet02-eth1", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet02:50", "name": "Tenet02-eth50", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}], "status": "up", "state": "enabled"}], "links": [], "services": ["l2vpn-ptp"]}'
2024-08-28T09:02:28.245063093Z DEBUG:sdx_controller.messaging.rpc_queue_consumer:MQ received message:b'{"name": "Tenet-OXP", "id": "urn:sdx:topology:tenet.ac.za", "version": 10, "timestamp": "2024-08-28T09:02:24Z", "model_version": "2.0.0", "nodes": [{"name": "Tenet02", "id": "urn:sdx:node:tenet.ac.za:Tenet02", "location": {"address": "Johanesburgo", "latitude": -26.0, "longitude": 28.0, "iso3166_2_lvl4": "ZA-GP", "private": []}, "ports": [{"id": "urn:sdx:port:tenet.ac.za:Tenet02:41", "name": "Tenet02-eth41", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "urn:sdx:port:sax.net:Sax02:41", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet02:1", "name": "Tenet02-eth1", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet02:50", "name": "Tenet02-eth50", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}], "status": "up", "state": "enabled"}, {"name": "Tenet03", "id": "urn:sdx:node:tenet.ac.za:Tenet03", "location": {"address": "EastLondon", "latitude": -33.0, "longitude": 27.0, "iso3166_2_lvl4": "ZA-EC", "private": []}, "ports": [{"id": "urn:sdx:port:tenet.ac.za:Tenet03:2", "name": "Tenet03-eth2", "node": "urn:sdx:node:tenet.ac.za:Tenet03", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet03:50", "name": "Tenet03-eth50", "node": "urn:sdx:node:tenet.ac.za:Tenet03", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}], "status": "up", "state": "enabled"}, {"name": "Tenet01", "id": "urn:sdx:node:tenet.ac.za:Tenet01", "location": {"address": "CapeTown", "latitude": -33.0, "longitude": 18.0, "iso3166_2_lvl4": "ZA-WC", "private": []}, "ports": [{"id": "urn:sdx:port:tenet.ac.za:Tenet01:41", "name": "Tenet01-eth41", "node": "urn:sdx:node:tenet.ac.za:Tenet01", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "urn:sdx:port:sax.net:Sax01:41", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet01:1", "name": "Tenet01-eth1", "node": "urn:sdx:node:tenet.ac.za:Tenet01", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet01:2", "name": "Tenet01-eth2", "node": "urn:sdx:node:tenet.ac.za:Tenet01", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet01:50", "name": "Tenet01-eth50", "node": "urn:sdx:node:tenet.ac.za:Tenet01", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}], "status": "up", "state": "enabled"}], "links": [{"name": "Tenet01/1_Tenet02/1", "id": "urn:sdx:link:tenet.ac.za:Tenet01/1_Tenet02/1", "ports": ["urn:sdx:port:tenet.ac.za:Tenet02:1", "urn:sdx:port:tenet.ac.za:Tenet01:1"], "type": "intra", "bandwidth": 10, "residual_bandwidth": 100, "latency": 0, "packet_loss": 0, "availability": 0, "status": "up", "state": "enabled", "private": ["packet_loss"]}, {"name": "Tenet01/2_Tenet03/2", "id": "urn:sdx:link:tenet.ac.za:Tenet01/2_Tenet03/2", "ports": ["urn:sdx:port:tenet.ac.za:Tenet03:2", "urn:sdx:port:tenet.ac.za:Tenet01:2"], "type": "intra", "bandwidth": 10, "residual_bandwidth": 100, "latency": 0, "packet_loss": 0, "availability": 0, "status": "up", "state": "enabled", "private": ["packet_loss"]}], "services": ["l2vpn-ptp"]}'
2024-08-28T09:02:28.245691969Z INFO:sdx_controller.handlers.lc_message_handler:MQ received message:b'{"name": "Tenet-OXP", "id": "urn:sdx:topology:tenet.ac.za", "version": 10, "timestamp": "2024-08-28T09:02:24Z", "model_version": "2.0.0", "nodes": [{"name": "Tenet02", "id": "urn:sdx:node:tenet.ac.za:Tenet02", "location": {"address": "Johanesburgo", "latitude": -26.0, "longitude": 28.0, "iso3166_2_lvl4": "ZA-GP", "private": []}, "ports": [{"id": "urn:sdx:port:tenet.ac.za:Tenet02:41", "name": "Tenet02-eth41", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "urn:sdx:port:sax.net:Sax02:41", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet02:1", "name": "Tenet02-eth1", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet02:50", "name": "Tenet02-eth50", "node": "urn:sdx:node:tenet.ac.za:Tenet02", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}], "status": "up", "state": "enabled"}, {"name": "Tenet03", "id": "urn:sdx:node:tenet.ac.za:Tenet03", "location": {"address": "EastLondon", "latitude": -33.0, "longitude": 27.0, "iso3166_2_lvl4": "ZA-EC", "private": []}, "ports": [{"id": "urn:sdx:port:tenet.ac.za:Tenet03:2", "name": "Tenet03-eth2", "node": "urn:sdx:node:tenet.ac.za:Tenet03", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet03:50", "name": "Tenet03-eth50", "node": "urn:sdx:node:tenet.ac.za:Tenet03", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}], "status": "up", "state": "enabled"}, {"name": "Tenet01", "id": "urn:sdx:node:tenet.ac.za:Tenet01", "location": {"address": "CapeTown", "latitude": -33.0, "longitude": 18.0, "iso3166_2_lvl4": "ZA-WC", "private": []}, "ports": [{"id": "urn:sdx:port:tenet.ac.za:Tenet01:41", "name": "Tenet01-eth41", "node": "urn:sdx:node:tenet.ac.za:Tenet01", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "urn:sdx:port:sax.net:Sax01:41", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet01:1", "name": "Tenet01-eth1", "node": "urn:sdx:node:tenet.ac.za:Tenet01", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet01:2", "name": "Tenet01-eth2", "node": "urn:sdx:node:tenet.ac.za:Tenet01", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}, {"id": "urn:sdx:port:tenet.ac.za:Tenet01:50", "name": "Tenet01-eth50", "node": "urn:sdx:node:tenet.ac.za:Tenet01", "type": "10GE", "status": "up", "state": "enabled", "mtu": 1500, "nni": "", "services": {"l2vpn-ptp": {"vlan_range": [[1, 4095]]}}, "private": ["status"]}], "status": "up", "state": "enabled"}], "links": [{"name": "Tenet01/1_Tenet02/1", "id": "urn:sdx:link:tenet.ac.za:Tenet01/1_Tenet02/1", "ports": ["urn:sdx:port:tenet.ac.za:Tenet02:1", "urn:sdx:port:tenet.ac.za:Tenet01:1"], "type": "intra", "bandwidth": 10, "residual_bandwidth": 100, "latency": 0, "packet_loss": 0, "availability": 0, "status": "up", "state": "enabled", "private": ["packet_loss"]}, {"name": "Tenet01/2_Tenet03/2", "id": "urn:sdx:link:tenet.ac.za:Tenet01/2_Tenet03/2", "ports": ["urn:sdx:port:tenet.ac.za:Tenet03:2", "urn:sdx:port:tenet.ac.za:Tenet01:2"], "type": "intra", "bandwidth": 10, "residual_bandwidth": 100, "latency": 0, "packet_loss": 0, "availability": 0, "status": "up", "state": "enabled", "private": ["packet_loss"]}], "services": ["l2vpn-ptp"]}'
```

As you can see above, the first topology event does not contains the full topology. Other nodes and links are added on subsequent events. This is normal behavior and can happen in real-world scenarios (beside tests like the one above). Since the `_update_vlan_tags_table_from_links ` is only called when the topology is added, rather than when is added *AND updated* the problem above can happen.

--- a/src/sdx_pce/topology/manager.py
+++ b/src/sdx_pce/topology/manager.py
@@ -33,11 +33,11 @@ class TopologyManager:
         # Mapping from topology ID to topology.
         self._topology_map = {}
 
-        # Mapping from port ID to link.
+        # Mapping from port ID to port.
         self._port_map = {}
 
-        # Mapping from port ID to node.
-        self._port_node_map = {}
+        # Mapping from port ID to link.
+        self._port_link_map = {}
 
         # Number of interdomain links we computed.
         self._num_interdomain_link = 0
@@ -92,22 +92,22 @@ class TopologyManager:
     def get_topology_map(self) -> dict:
         return self._topology_map
 
-    def get_port_map(self) -> Mapping[str, dict]:
+    def get_port_link_map(self) -> Mapping[str, dict]:
         """
         Return a mapping between port IDs and links.
         """
-        return self._port_map
+        return self._port_link_map
 
-    def get_port_node_map(self) -> Mapping[str, dict]:
+    def get_port_map(self) -> Mapping[str, dict]:
         """
-        Return a mapping between port IDs and nodes.
+        Return a mapping between port IDs and ports.
         """
-        return self._port_node_map
+        return self._port_map
 
     def clear_topology(self):
         self._topology = None
         self._topology_map = {}
-        self._port_map = {}
+        self._port_link_map = {}
 
     def add_topology(self, data):
         topology = TopologyHandler().import_topology_data(data)
@@ -124,7 +124,7 @@ class TopologyManager:
             # links = topology.links
             # for link in links:
             #    for port in link.ports:
-            #        self._port_map[port["id"]] = link
+            #        self._port_link_map[port["id"]] = link
         else:
             # check the inter-domain links first.
             interdomain_ports = self.inter_domain_check(topology)
@@ -150,13 +150,13 @@ class TopologyManager:
         for link in links:
             for port in link.ports:
                 port_id = port if isinstance(port, str) else port["id"]
-                self._port_map[port_id] = link
+                self._port_link_map[port_id] = link
 
         # Addding to the port node
         nodes = topology.nodes
         for node in nodes:
             for port in node.ports:
-                self._port_node_map[port.id] = node
+                self._port_map[port.id] = port
 
         # inter-domain links
         self.add_inter_domain_links(topology, interdomain_ports)
@@ -198,7 +198,7 @@ class TopologyManager:
         """
         for port in link.ports:
             port_id = port if isinstance(port, str) else port["id"]
-            if port_id not in self._port_map:
+            if port_id not in self._port_link_map:
                 return True
         return False
 
@@ -235,7 +235,7 @@ class TopologyManager:
                 self._topology.remove_link(link.id)
                 for port in link.ports:
                     port_id = port if isinstance(port, str) else port["id"]
-                    self._port_map.pop(port_id)
+                    self._port_link_map.pop(port_id)
 
         # Check the inter-domain links first.
         interdomain_ports = self.inter_domain_check(topology)
@@ -256,7 +256,7 @@ class TopologyManager:
         # Update the port node map
         for node in topology.nodes:
             for port in node.ports:
-                self._port_node_map[port.id] = node
+                self._port_map[port.id] = port
 
         self.update_version(True)
         self.update_timestamp()
@@ -303,14 +303,14 @@ class TopologyManager:
         for port_id in interdomain_port_dict:
             # print("interdomain_port:")
             # print(port_id)
-            for existing_port, existing_link in self._port_map.items():
+            for existing_port, existing_link in self._port_link_map.items():
                 # print(existing_port)
                 if port_id == existing_port:
                     # print("Interdomain port:" + port_id)
                     # remove redundant link between two domains
                     self._topology.remove_link(existing_link.id)
                     interdomain_port_ids.append(port_id)
-            self._port_map[port_id] = interdomain_port_dict[port_id]
+            self._port_link_map[port_id] = interdomain_port_dict[port_id]
 
         # count for inter-domain links according to topo spec 2.0.x
         for node in topology.nodes:
@@ -358,15 +358,11 @@ class TopologyManager:
     def add_inter_domain_links(self, topology, interdomain_ports):
         """Add inter-domain links (whenever possible)."""
         for port in interdomain_ports:
-            other_node = self._port_node_map.get(port.nni)
-            other_ports = other_node.ports if other_node else []
-            for other_port in other_ports:
-                if other_port.id == port.nni and other_port.nni == port.id:
-                    break
-            else:
+            other_port = self._port_map.get(port.nni)
+            if not other_port or other_port.nni != port.id:
                 self._logger.warning(
                     "Interdomain link not added now - didnt find other port:"
-                    f" port={port.id} other_port={port.nni}"
+                    f" port={port.id} other_port={port.nni} ({other_port})"
                 )
                 continue
             self.create_update_interdomain_link(port, other_port)

--- a/src/sdx_pce/topology/temanager.py
+++ b/src/sdx_pce/topology/temanager.py
@@ -58,7 +58,7 @@ class TEManager:
             self.graph = self.generate_graph_te()
             self._update_vlan_tags_table(
                 domain_name=topology_data.get("id"),
-                port_map=self.topology_manager.get_port_node_map(),
+                port_map=self.topology_manager.get_port_map(),
             )
         else:
             self.graph = None
@@ -77,7 +77,7 @@ class TEManager:
         # attached to links.
         self._update_vlan_tags_table(
             domain_name=topology_data.get("id"),
-            port_map=self.topology_manager.get_port_node_map(),
+            port_map=self.topology_manager.get_port_map(),
         )
 
     def update_topology(self, topology_data: dict):
@@ -88,27 +88,21 @@ class TEManager:
         """
         self.topology_manager.update_topology(topology_data)
 
+        # Update vlan_tags_table in a non-disruptive way. Previous concerned
+        # still applies:
         # TODO: careful here when updating VLAN tags table -- what do
         # we do when an in use VLAN tag becomes invalid in the update?
         # See https://github.com/atlanticwave-sdx/pce/issues/123
-        #
-        # self._update_vlan_tags_table_from_links(
-        #     domain_name=topology_data.get("id"),
-        #     port_list=self.topology_manager.port_list,
-        # )
+        self._update_vlan_tags_table(
+            domain_name=topology_data.get("id"),
+            port_map=self.topology_manager.get_port_map(),
+        )
 
     def get_topology_map(self) -> dict:
         """
         Get {topology_id: topology, ..} map.
         """
         return self.topology_manager.get_topology_map()
-
-    def get_port_services_label_range(self, port: dict) -> List[str]:
-        vlan_range = None
-        services = port.get("services")
-        if services and services.get("l2vpn-ptp"):
-            vlan_range = services.get("l2vpn-ptp").get("vlan_range")
-        return vlan_range
 
     def get_port_obj_services_label_range(self, port: Port) -> List[str]:
         vlan_range = None
@@ -123,95 +117,36 @@ class TEManager:
 
     def _update_vlan_tags_table(self, domain_name: str, port_map: dict):
         """
-        Update VLAN tags table.
+        Update VLAN tags table in a non-disruptive way, meaning: only add new
+        VLANs to the table. Removed VLANs will need further discussion (see
+        https://github.com/atlanticwave-sdx/pce/issues/123)
         """
-        self._vlan_tags_table[domain_name] = {}
+        self._vlan_tags_table.setdefault(domain_name, {})
 
-        for port_id, node in port_map.items():
-            # TODO: port here seems to be a dict, not sdx_datamodel.models.Port
-            for port in node.ports:
-                # Collect all port IDs in this link.  Each link should
-                # have two ports.
-                node_port_ids = [x.id for x in node.ports]
+        for port_id, port in port_map.items():
+            # Get the label range for this port: either from the
+            # port itself (v1), or from the services attached to it (v2).
+            label_range = self.get_port_obj_services_label_range(port)
+            if label_range is None:
+                label_range = port.vlan_range
 
-                # Do some error checks.
-                node_port_count = len(node_port_ids)
+            # TODO: why is label_range sometimes None, and what to
+            # do when that happens?
+            if label_range is None:
+                self._logger.info(f"label_range on {port.id} is None")
+                continue
 
-                if node_port_count < 1:
-                    raise ValidationError(
-                        f"Node has {node_port_count} ports, not greater than 0"
-                    )
+            # label_range is of the form ['100-200', '1000']; let
+            # us expand it.  Would have been ideal if this was
+            # already in some parsed form, but it is not, so this
+            # is a work-around.
+            all_labels = self._expand_label_range(label_range)
 
-                if port_id not in node_port_ids:
-                    raise ValidationError(f"port {port_id} not in {node_port_ids}")
-
-                # Get the label range for this port: either from the
-                # port itself (v1), or from the services attached to it (v2).
-                label_range = self.get_port_obj_services_label_range(port)
-                if label_range is None:
-                    label_range = port.vlan_range
-
-                # TODO: why is label_range sometimes None, and what to
-                # do when that happens?
-                if label_range is None:
-                    self._logger.info(f"label_range on {port.id} is None")
-                    continue
-
-                # label_range is of the form ['100-200', '1000']; let
-                # us expand it.  Would have been ideal if this was
-                # already in some parsed form, but it is not, so this
-                # is a work-around.
-                all_labels = self._expand_label_range(label_range)
-
-                # Make a map like: `{label1: UNUSED_VLAN, label2: UNUSED_VLAN,...}`
-                labels_available = {label: UNUSED_VLAN for label in all_labels}
-
-                self._vlan_tags_table[domain_name][port_id] = labels_available
-
-    def _update_vlan_tags_table_from_links(self, domain_name: str, port_map: dict):
-        """
-        Update VLAN tags table.
-        """
-        self._vlan_tags_table[domain_name] = {}
-
-        for port_id, link in port_map.items():
-            # TODO: port here seems to be a dict, not sdx_datamodel.models.Port
-            for port in link.ports:
-                # Collect all port IDs in this link.  Each link should
-                # have two ports.
-                link_port_ids = [x["id"] for x in link.ports]
-
-                # Do some error checks.
-                link_port_count = len(link_port_ids)
-
-                if link_port_count != 2:
-                    raise ValidationError(f"Link has {link_port_count} ports, not 2")
-
-                if port_id not in link_port_ids:
-                    raise ValidationError(f"port {port_id} not in {link_port_ids}")
-
-                # Get the label range for this port: either from the
-                # port itself (v1), or from the services attached to it (v2).
-                label_range = self.get_port_services_label_range(port)
-                if label_range is None:
-                    label_range = port.get("label_range")
-
-                # TODO: why is label_range sometimes None, and what to
-                # do when that happens?
-                if label_range is None:
-                    self._logger.info(f"label_range on {port['id']} is None")
-                    continue
-
-                # label_range is of the form ['100-200', '1000']; let
-                # us expand it.  Would have been ideal if this was
-                # already in some parsed form, but it is not, so this
-                # is a work-around.
-                all_labels = self._expand_label_range(label_range)
-
-                # Make a map like: `{label1: UNUSED_VLAN, label2: UNUSED_VLAN,...}`
-                labels_available = {label: UNUSED_VLAN for label in all_labels}
-
-                self._vlan_tags_table[domain_name][port_id] = labels_available
+            self._vlan_tags_table[domain_name].setdefault(port_id, {})
+            for label in all_labels:
+                self._vlan_tags_table[domain_name][port_id].setdefault(
+                    label, UNUSED_VLAN
+                )
 
     def _expand_label_range(self, label_range: []) -> List[int]:
         """
@@ -543,7 +478,7 @@ class TEManager:
                 if (
                     not request_format_is_tm
                     and connection_request["ingress_port"]["id"]
-                    not in self.topology_manager.get_port_map()
+                    not in self.topology_manager.get_port_link_map()
                 ):
                     self._logger.warning(
                         f"Port {connection_request['ingress_port']['id']} not found in port map, it's a user port"
@@ -564,7 +499,7 @@ class TEManager:
                     not request_format_is_tm
                     and same_domain_port_flag
                     and connection_request["egress_port"]["id"]
-                    not in self.topology_manager.get_port_map()
+                    not in self.topology_manager.get_port_link_map()
                 ):
                     self._logger.warning(
                         f"Port {connection_request['egress_port']['id']} not found in port map, it's a user port"
@@ -585,7 +520,7 @@ class TEManager:
                 if (
                     not request_format_is_tm
                     and connection_request["egress_port"]["id"]
-                    not in self.topology_manager.get_port_map()
+                    not in self.topology_manager.get_port_link_map()
                 ):
                     self._logger.warning(
                         f"Port {connection_request['egress_port']['id']} not found in port map, it's a user port"
@@ -764,7 +699,7 @@ class TEManager:
             # or (2) uses the OXP vlan if (2.1) not provided or provided (2.2) is not in the vlan range in the topology port.
             # And we do't allow user specified vlan on a OXP port.
             if (
-                ingress_port_id not in self.topology_manager.get_port_map()
+                ingress_port_id not in self.topology_manager.get_port_link_map()
                 and ingress_vlan is None
             ):
                 self._logger.warning(
@@ -773,7 +708,7 @@ class TEManager:
                 ingress_vlan = egress_vlan
 
             if (
-                egress_port_id not in self.topology_manager.get_port_map()
+                egress_port_id not in self.topology_manager.get_port_link_map()
                 and egress_vlan is None
             ):
                 self._logger.warning(

